File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/runtime/RuntimeOptions.java
package org.maplibre.nativeffi.runtime;

import java.util.OptionalLong;

/** Mutable descriptor used when creating a {@link RuntimeHandle}. */
public final class RuntimeOptions {
  private String assetPath;
  private String cachePath;
  private Long maximumCacheSize;

  public String assetPath() {
    return assetPath;
  }

  public RuntimeOptions assetPath(String assetPath) {
    this.assetPath = assetPath;
    return this;
  }

  public RuntimeOptions clearAssetPath() {
    assetPath = null;
    return this;
  }

  public String cachePath() {
    return cachePath;
  }

  public RuntimeOptions cachePath(String cachePath) {
    this.cachePath = cachePath;
    return this;
  }

  public RuntimeOptions clearCachePath() {
    cachePath = null;
    return this;
  }

  public OptionalLong maximumCacheSize() {
    return maximumCacheSize == null ? OptionalLong.empty() : OptionalLong.of(maximumCacheSize);
  }

  public boolean hasMaximumCacheSize() {
    return maximumCacheSize != null;
  }

  public RuntimeOptions maximumCacheSize(long maximumCacheSize) {
    this.maximumCacheSize = maximumCacheSize;
    return this;
  }

  public RuntimeOptions clearMaximumCacheSize() {
    maximumCacheSize = null;
    return this;
  }
}

File: bindings/kotlin/src/commonMain/kotlin/org/maplibre/nativeffi/runtime/RuntimeOptions.kt
package org.maplibre.nativeffi.runtime

/**
 * Mutable descriptor used when creating a [RuntimeHandle].
 *
 * Compares and hashes by field value; [copy] returns an independent instance. Keep an instance
 * unmodified while it is a key in a hash-based collection.
 */
public class RuntimeOptions {
  public var assetPath: String? = null

  public var cachePath: String? = null

  /**
   * Runtime-originated event types this runtime queues, the native library's default until a host
   * narrows it. That default selects every runtime-originated type the library reports, including a
   * type this version does not name, whose events reach a host as unknown event and payload
   * domains.
   *
   * [RuntimeHandle.create] fails with [org.maplibre.nativeffi.error.InvalidArgumentException] on a
   * bit the native library does not define.
   */
  public var eventMask: RuntimeEventMask = defaultRuntimeEventMask()

  /** Returns an independent copy of this descriptor with [block] applied to the copy. */
  public fun copy(block: RuntimeOptions.() -> Unit = {}): RuntimeOptions =
    RuntimeOptions()
      .also {
        it.assetPath = assetPath
        it.cachePath = cachePath
        it.eventMask = eventMask
      }
      .apply(block)

  private val fields: List<Any?>
    get() = listOf(assetPath, cachePath, eventMask)

  override fun equals(other: Any?): Boolean = other is RuntimeOptions && fields == other.fields

  override fun hashCode(): Int = fields.hashCode()
}
